# Đồ án thực hành: XÂY DỰNG VÀ KHAI THÁC KDL
- Môn: **HTTT Phục vụ trí tuệ kinh doanh**
- Lớp: **19HTTT2**
- Yêu cầu: **Xây dựng và phân tích dữ liệu về Covid-19 trong các năm 2020 - 2022**
- Thông tin nhóm:

|STT | MSSV | Họ tên |
| :---: | :---: | :--- |
| 1 | 19127304 | Trần Khải Trúc |
| 2 | 19127474 | Lương Kiếm Minh | 
| 3 | 19127649 | Tô Thanh Tuấn | 

In [163]:
import pandas as pd

## 1. Chuẩn bị dữ liệu `Cases Report`

### Đọc và xem bảng

In [164]:
# Đọc dữ liệu
f_path = 'Datasource/Cases Report.csv'
case_report_df = pd.read_csv(f_path)

# Chuẩn hóa tên cột
header_list = case_report_df.columns.str.title()
case_report_df.columns = [header.replace(' ','_').upper() for header in header_list]

# Xem thử dữ liệu
case_report_df.head()

OUTCOME  AGE  GENDER                      REPORTING_PHU SPECIMENDATE  \
0  Resolved  80s  FEMALE              Toronto Public Health    9/24/2020   
1  Resolved  <20    MALE                 Peel Public Health   12/24/2020   
2  Resolved  20s    MALE              Lambton Public Health    9/23/2020   
3  Resolved  20s    MALE  Region of Waterloo, Public Health    9/20/2020   
4  Resolved  20s  FEMALE               Ottawa Public Health    9/21/2020   

  CASEREPORTED_DATE       PHUCITY TESTREPORTED_DATE CASEACQUISITION_INFO  \
0          2/6/2021       Toronto          2/6/2021                   OB   
1          2/1/2021   Mississauga          2/1/2021                   CC   
2         1/15/2021  Point Edward         1/14/2021                   CC   
3        12/29/2020      Waterloo         9/23/2020    NO KNOWN EPI LINK   
4        12/24/2020        Ottawa         9/22/2020                   CC   

  ACCURATEEPISODE_DT                     PHU_ADDRESS  \
0          1/26/2021  277 Victoria Street, 5th Floor   
1          1/29/2021          7120 Hurontario Street   
2          1/10/2021              160 Exmouth Street   
3          9/20/2020          99 Regina Street South   
4          9/21/2020         100 Constellation Drive   

                                         PHU_WEBSITE OUTBREAKRELATED  \
0  www.toronto.ca/community-people/health-wellnes...             Yes   
1                          www.peelregion.ca/health/             NaN   
2                            www.lambtonhealth.on.ca             Yes   
3                            www.regionofwaterloo.ca             NaN   
4                          www.ottawapublichealth.ca             NaN   

   PHU_LATITUDE  PHU_LONGITUDE PHU_POSTAL_CODE  
0     43.656591     -79.379358         M5B 1W2  
1     43.647471     -79.708893         L5W 1N4  
2     42.986416     -82.404808         N7T 7Z6  
3     43.462876     -80.520913         N2J 4V3  
4     45.345665     -75.763912         K2G 6J8

### Xem thông tin số dòng và các cột của bảng

In [165]:
case_report_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28500 entries, 0 to 28499
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   OUTCOME               28500 non-null  object 
 1   AGE                   28500 non-null  object 
 2   GENDER                28500 non-null  object 
 3   REPORTING_PHU         28500 non-null  object 
 4   SPECIMENDATE          28292 non-null  object 
 5   CASEREPORTED_DATE     28500 non-null  object 
 6   PHUCITY               28500 non-null  object 
 7   TESTREPORTED_DATE     28082 non-null  object 
 8   CASEACQUISITION_INFO  28500 non-null  object 
 9   ACCURATEEPISODE_DT    28500 non-null  object 
 10  PHU_ADDRESS           28500 non-null  object 
 11  PHU_WEBSITE           28500 non-null  object 
 12  OUTBREAKRELATED       8710 non-null   object 
 13  PHU_LATITUDE          28500 non-null  float64
 14  PHU_LONGITUDE         28500 non-null  float64
 15  PHU_POSTAL_CODE    

$\rightarrow$ **Kết quả:** Có tổng cộng 28500 dòng dữ liệu. Các cột `SpecimenDate`, `TestReportedDate` và `OutbreakRelated` còn khuyết dữ liệu.

### Kiểm tra dữ liệu bị duplicated và xử lí

In [166]:
# Kiểm tra các dòng bị trùng lắp
case_report_df[case_report_df.duplicated()]

OUTCOME  AGE  GENDER                       REPORTING_PHU SPECIMENDATE  \
105    Resolved  <20  FEMALE     Hamilton Public Health Services    10/6/2020   
115    Resolved  20s  FEMALE                  Peel Public Health    10/3/2020   
134    Resolved  20s    MALE                  Peel Public Health    10/3/2020   
162    Resolved  30s  FEMALE                  Peel Public Health    10/7/2020   
197    Resolved  <20  FEMALE     Hamilton Public Health Services    10/6/2020   
...         ...  ...     ...                                 ...          ...   
27070  Resolved  30s  FEMALE                  Peel Public Health    3/28/2020   
27102  Resolved  40s    MALE  York Region Public Health Services    3/26/2020   
27509  Resolved  20s  FEMALE                  Peel Public Health    3/25/2020   
27720  Resolved  60s  FEMALE                  Peel Public Health    3/21/2020   
27786  Resolved  30s    MALE                  Peel Public Health    3/20/2020   

      CASEREPORTED_DATE      PHUCITY TESTREPORTED_DATE CASEACQUISITION_INFO  \
105           10/9/2020     Hamilton         10/9/2020                   CC   
115           10/9/2020  Mississauga         10/9/2020                   CC   
134           10/9/2020  Mississauga         10/9/2020                   CC   
162           10/9/2020  Mississauga         10/9/2020                   CC   
197           10/8/2020     Hamilton         10/8/2020                   OB   
...                 ...          ...               ...                  ...   
27070         3/30/2020  Mississauga         3/30/2020                   CC   
27102         3/30/2020    Newmarket         3/30/2020    NO KNOWN EPI LINK   
27509         3/28/2020  Mississauga         3/28/2020                   CC   
27720         3/27/2020  Mississauga         3/27/2020                   CC   
27786         3/26/2020  Mississauga         3/26/2020    NO KNOWN EPI LINK   

      ACCURATEEPISODE_DT                   PHU_ADDRESS  \
105            10/6/2020  110 King St. West, 2nd Floor   
115            10/3/2020        7120 Hurontario Street   
134            10/3/2020        7120 Hurontario Street   
162            10/7/2020        7120 Hurontario Street   
197            10/4/2020  110 King St. West, 2nd Floor   
...                  ...                           ...   
27070          3/23/2020        7120 Hurontario Street   
27102          3/19/2020            17250 Yonge Street   
27509          3/24/2020        7120 Hurontario Street   
27720          3/14/2020        7120 Hurontario Street   
27786          3/17/2020        7120 Hurontario Street   

                                    PHU_WEBSITE OUTBREAKRELATED  PHU_LATITUDE  \
105                www.hamilton.ca/publichealth             NaN     43.257631   
115                   www.peelregion.ca/health/             NaN     43.647471   
134                   www.peelregion.ca/health/             NaN     43.647471   
162                   www.peelregion.ca/health/             NaN     43.647471   
197                www.hamilton.ca/publichealth             Yes     43.257631   
...                                         ...             ...           ...   
27070                 www.peelregion.ca/health/             NaN     43.647471   
27102  www.york.ca/wps/portal/yorkhome/health/              NaN     44.048023   
27509                 www.peelregion.ca/health/             NaN     43.647471   
27720                 www.peelregion.ca/health/             NaN     43.647471   
27786                 www.peelregion.ca/health/             NaN     43.647471   

       PHU_LONGITUDE PHU_POSTAL_CODE  
105       -79.871341         L8P 4S6  
115       -79.708893         L5W 1N4  
134       -79.708893         L5W 1N4  
162       -79.708893         L5W 1N4  
197       -79.871341         L8P 4S6  
...              ...             ...  
27070     -79.708893         L5W 1N4  
27102     -79.480239         L3Y 6Z1  
27509     -79.708893         L5W 1N4  
27720     -79.708893      

$\rightarrow$ **Kết quả:** Có tổng cộng 1528 dòng dữ liệu bị trùng lắp.

In [167]:
# Xử lí trùng lắp bằng cách xóa dòng
case_report_df = case_report_df.drop_duplicates()

In [168]:
# Kiểm tra xem các dòng bị trùng lắp đã được xóa chưa
case_report_df[case_report_df.duplicated()]

Empty DataFrame
Columns: [OUTCOME, AGE, GENDER, REPORTING_PHU, SPECIMENDATE, CASEREPORTED_DATE, PHUCITY, TESTREPORTED_DATE, CASEACQUISITION_INFO, ACCURATEEPISODE_DT, PHU_ADDRESS, PHU_WEBSITE, OUTBREAKRELATED, PHU_LATITUDE, PHU_LONGITUDE, PHU_POSTAL_CODE]
Index: []

$\rightarrow$ **Kết quả:** Bảng `Cases Report` đã hết các dòng dữ liệu bị trùng lắp.

### Xuất dữ liệu hoàn chỉnh ra file source

In [169]:
case_report_df.to_csv('Datasource_Cleaned/Case_Report.csv')

## 2. Chuẩn bị dữ liệu `Compiled_COVID-19_Case_Details_(Canada)`

### Đọc và xem bảng

In [170]:
# Đọc dữ liệu
f_path = 'Datasource/Compiled_COVID-19_Case_Details_(Canada).csv'
case_detail_df = pd.read_csv(f_path)

# Chuẩn hóa tên cột
header_list = case_detail_df.columns.str.title()
case_detail_df.columns = [header.upper() for header in header_list]

# Xem thử dữ liệu
case_detail_df.head()

OBJECTID  ROW_ID           DATE_REPORTED  \
0         1     500  2020/03/23 12:00:00+00   
1         2     501  2020/04/02 12:00:00+00   
2         3     502  2020/03/25 12:00:00+00   
3         4     503  2020/03/28 12:00:00+00   
4         5     504  2020/03/30 12:00:00+00   

                                       HEALTH_REGION AGE_GROUP  GENDER  \
0                                 Peel Public Health     20-29  Female   
1                                 Peel Public Health     40-49  Female   
2  Kingston, Frontenac and Lennox & Addington Pub...     30-39    Male   
3  Kingston, Frontenac and Lennox & Addington Pub...       <20  Female   
4                                 Peel Public Health     50-59    Male   

        EXPOSURE CASE_STATUS PROVINCE  
0  Close Contact   Recovered  Ontario  
1  Close Contact   Recovered  Ontario  
2  Close Contact   Recovered  Ontario  
3  Close Contact   Recovered  Ontario  
4  Close Contact   Recovered  Ontario

### Xem thông tin số dòng và các cột của bảng

In [171]:
case_detail_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype 
---  ------         --------------    ----- 
 0   OBJECTID       1048575 non-null  int64 
 1   ROW_ID         1048575 non-null  int64 
 2   DATE_REPORTED  1048575 non-null  object
 3   HEALTH_REGION  1048575 non-null  object
 4   AGE_GROUP      1048575 non-null  object
 5   GENDER         1048575 non-null  object
 6   EXPOSURE       1048575 non-null  object
 7   CASE_STATUS    1048575 non-null  object
 8   PROVINCE       1048575 non-null  object
dtypes: int64(2), object(7)
memory usage: 72.0+ MB


$\rightarrow$ **Kết quả:** Có tổng cộng 1048575 dòng dữ liệu và không có cột nào bị khuyết dữ liệu.

### Kiểm tra dữ liệu bị duplicated và xử lí

In [172]:
# Kiểm tra các dòng bị trùng lắp
case_detail_df[case_detail_df.duplicated()]

Empty DataFrame
Columns: [OBJECTID, ROW_ID, DATE_REPORTED, HEALTH_REGION, AGE_GROUP, GENDER, EXPOSURE, CASE_STATUS, PROVINCE]
Index: []

$\rightarrow$ **Kết quả:** Không có dòng dữ liệu bị duplicated nên không cần xử lí.

### Xuất dữ liệu hoàn chỉnh ra file source

In [173]:
case_detail_df.to_csv('Datasource_Cleaned/Case_Detail.csv')

## 3. Chuẩn bị dữ liệu `ongoing_outbreaks_phu`

### Đọc và xem bảng

In [174]:
# Đọc dữ liệu
f_path = 'Datasource/ongoing_outbreaks_phu.csv'
ongoing_outbreak_df = pd.read_csv(f_path)

# Chuẩn hóa tên cột
header_list = ongoing_outbreak_df.columns.str.title()
ongoing_outbreak_df.columns = [header.upper() for header in header_list]

# Xem thử dữ liệu
ongoing_outbreak_df.head()

DATE  PHU_NUM     OUTBREAK_GROUP  NUMBER_ONGOING_OUTBREAKS
0  11/1/2020     2227  1 Congregate Care                         2
1  11/1/2020     2227        4 Workplace                         1
2  11/1/2020     2240        4 Workplace                         1
3  11/1/2020     2240     5 Recreational                         1
4  11/1/2020     2240    6 Other/Unknown                         1

### Xem thông tin số dòng và các cột của bảng

In [175]:
ongoing_outbreak_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56987 entries, 0 to 56986
Data columns (total 4 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   DATE                      56987 non-null  object
 1   PHU_NUM                   56987 non-null  int64 
 2   OUTBREAK_GROUP            56987 non-null  object
 3   NUMBER_ONGOING_OUTBREAKS  56987 non-null  int64 
dtypes: int64(2), object(2)
memory usage: 1.7+ MB


$\rightarrow$ **Kết quả:** Có tổng cộng 56987 dòng dữ liệu và không có cột nào bị khuyết dữ liệu.

### Kiểm tra dữ liệu bị duplicated và xử lí

In [176]:
# Kiểm tra các dòng bị trùng lắp
ongoing_outbreak_df[ongoing_outbreak_df.duplicated()]

Empty DataFrame
Columns: [DATE, PHU_NUM, OUTBREAK_GROUP, NUMBER_ONGOING_OUTBREAKS]
Index: []

$\rightarrow$ **Kết quả:** Không có dòng dữ liệu bị duplicated nên không cần xử lí.

### Xuất dữ liệu hoàn chỉnh ra file source

In [177]:
ongoing_outbreak_df.to_csv('Datasource_Cleaned/Ongoing_Outbreak.csv')

## 4. Chuẩn bị dữ liệu `Public health unit`

### Đọc và xem bảng

In [178]:
# Đọc dữ liệu
f_path = 'Datasource/Public health unit.xlsx'
phu_df = pd.read_excel(f_path)

# Chuẩn hóa tên cột
header_list = phu_df.columns.str.title()
phu_df.columns = [header.upper() for header in header_list]

# Xem thử dữ liệu
phu_df.head()

PHU_ID                             REPORTING_PHU  \
0    2244              Middlesex-London Health Unit   
1    2236           Halton Region Health Department   
2    3895                     Toronto Public Health   
3    5183          Huron Perth District Health Unit   
4    2266  Wellington-Dufferin-Guelph Public Health   

            REPORTING_PHU_ADDRESS REPORTING_PHU_CITY  \
0                  50 King Street             London   
1                1151 Bronte Road           Oakville   
2  277 Victoria Street, 5th Floor            Toronto   
3            653 West Gore Street          Stratford   
4             160 Chancellors Way             Guelph   

  REPORTING_PHU_POSTAL_CODE  \
0                   N6A 5L7   
1                   L6M 3Ll   
2                   M5B 1W2   
3                   N5A 1L4   
4                   N1G 0E1   

                               REPORTING_PHU_WEBSITE  REPORTING_PHU_LATITUDE  \
0                                 www.healthunit.com              4299571559   
1         www.halton.ca/For-Residents/Public-Health/              4351209709   
2  www.toronto.ca/community-people/health-wellnes...              4372584482   
3                                        www.hpph.ca              4360051407   
4                             www.wdgpublichealth.ca              4385364154   

   REPORTING_PHU_LONGITUDE  
0               -814569832  
1              -7989050809  
2              -7938955775  
3              -8128148453  
4              -8037208498

### Xem thông tin số dòng và các cột của bảng

In [179]:
phu_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   PHU_ID                     35 non-null     int64 
 1   REPORTING_PHU              35 non-null     object
 2   REPORTING_PHU_ADDRESS      35 non-null     object
 3   REPORTING_PHU_CITY         35 non-null     object
 4   REPORTING_PHU_POSTAL_CODE  35 non-null     object
 5   REPORTING_PHU_WEBSITE      35 non-null     object
 6   REPORTING_PHU_LATITUDE     35 non-null     int64 
 7   REPORTING_PHU_LONGITUDE    35 non-null     int64 
dtypes: int64(3), object(5)
memory usage: 2.3+ KB


$\rightarrow$ **Kết quả:** Có tổng cộng 35 dòng dữ liệu và không có cột nào bị khuyết dữ liệu.

### Kiểm tra dữ liệu bị duplicated và xử lí

In [180]:
# Kiểm tra các dòng bị trùng lắp
phu_df[phu_df.duplicated()]

PHU_ID                 REPORTING_PHU REPORTING_PHU_ADDRESS  \
34    2244  Middlesex-London Health Unit        50 King Street   

   REPORTING_PHU_CITY REPORTING_PHU_POSTAL_CODE REPORTING_PHU_WEBSITE  \
34             London                   N6A 5L7    www.healthunit.com   

    REPORTING_PHU_LATITUDE  REPORTING_PHU_LONGITUDE  
34              4299571559               -814569832

$\rightarrow$ **Kết quả:** Có tổng cộng 1 dòng dữ liệu bị trùng lắp.

In [181]:
# Xử lí trùng lắp bằng cách xóa dòng
phu_df = phu_df.drop_duplicates()

In [182]:
# Kiểm tra xem các dòng bị trùng lắp đã được xóa chưa
phu_df[phu_df.duplicated()]

Empty DataFrame
Columns: [PHU_ID, REPORTING_PHU, REPORTING_PHU_ADDRESS, REPORTING_PHU_CITY, REPORTING_PHU_POSTAL_CODE, REPORTING_PHU_WEBSITE, REPORTING_PHU_LATITUDE, REPORTING_PHU_LONGITUDE]
Index: []

$\rightarrow$ **Kết quả:** Bảng `Public Health Unit` đã hết các dòng dữ liệu bị trùng lắp.

### Xuất dữ liệu hoàn chỉnh ra file source

In [183]:
phu_df.to_csv('Datasource_Cleaned/Public_Health_Unit.csv')

## 5. Chuẩn bị dữ liệu `Public Health Units GROUP`

### Đọc và xem bảng

In [184]:
# Đọc dữ liệu
f_path = 'Datasource/Public Health Units GROUP.xlsx'
phu_group_df = pd.read_excel(f_path)

# Chuẩn hóa tên cột
header_list = phu_group_df.columns.str.title()
phu_group_df.columns = [header.upper() for header in header_list]

# Xem thử dữ liệu
phu_group_df.head()

PHU_GROUP       PHU_CITY  \
0       Central        Toronto   
1       Central      Newmarket   
2  Central East  Durham Region   
3  Central East      Port Hope   
4  Central East   Peterborough   

                                          PHU_REGION  
0                              Toronto Public Health  
1                 York Region Public Health Services  
2                    Durham Region Health Department  
3  Haliburton, Kawartha, Pine Ridge District Heal...  
4                         Peterborough Public Health

### Xem thông tin số dòng và các cột của bảng

In [185]:
phu_group_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   PHU_GROUP   34 non-null     object
 1   PHU_CITY    34 non-null     object
 2   PHU_REGION  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


$\rightarrow$ **Kết quả:** Có tổng cộng 34 dòng dữ liệu và không có cột nào bị khuyết dữ liệu.

### Kiểm tra dữ liệu bị duplicated và xử lí

In [186]:
# Kiểm tra các dòng bị trùng lắp
phu_group_df[phu_group_df.duplicated()]

Empty DataFrame
Columns: [PHU_GROUP, PHU_CITY, PHU_REGION]
Index: []

$\rightarrow$ **Kết quả:** Không có dòng dữ liệu bị duplicated nên không cần xử lí.

### Xuất dữ liệu hoàn chỉnh ra file source

In [187]:
phu_group_df.to_csv('Datasource_Cleaned/Public_Health_Units_GROUP.csv')

## 6. Chuẩn bị dữ liệu `vaccines_by_age_phu`

### Đọc và xem bảng

In [189]:
# Đọc dữ liệu
f_path = 'Datasource/vaccines_by_age_phu.csv'
vaccines_df = pd.read_csv(f_path)

# Chuẩn hóa tên cột
header_list = vaccines_df.columns.str.title()
vaccines_df.columns = [header.replace(' ','_').upper() for header in header_list]

# Xem thử dữ liệu
vaccines_df.head()

DATE  PHU_ID  AGEGROUP  AT_LEAST_ONE_DOSE_CUMULATIVE  \
0  7/26/2021    2226  12-17yrs                        3843.0   
1  7/26/2021    2227  12-17yrs                        6306.0   
2  7/26/2021    2230  12-17yrs                       34025.0   
3  7/26/2021    2233  12-17yrs                        5533.0   
4  7/26/2021    2234  12-17yrs                        3125.0   

   SECOND_DOSE_CUMULATIVE  FULLY_VACCINATED_CUMULATIVE  THIRD_DOSE_CUMULATIVE  
0                  1890.0                          NaN                    NaN  
1                  3567.0                          NaN                    NaN  
2                 16953.0                          NaN                    NaN  
3                  3627.0                          NaN                    NaN  
4                  1711.0                          NaN                    NaN

### Xem thông tin số dòng và các cột của bảng

In [190]:
vaccines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 171182 entries, 0 to 171181
Data columns (total 7 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   DATE                          171182 non-null  object 
 1   PHU_ID                        171182 non-null  int64  
 2   AGEGROUP                      171182 non-null  object 
 3   AT_LEAST_ONE_DOSE_CUMULATIVE  170971 non-null  float64
 4   SECOND_DOSE_CUMULATIVE        48642 non-null   float64
 5   FULLY_VACCINATED_CUMULATIVE   121912 non-null  float64
 6   THIRD_DOSE_CUMULATIVE         106015 non-null  float64
dtypes: float64(4), int64(1), object(2)
memory usage: 9.1+ MB


$\rightarrow$ **Kết quả:** Có tổng cộng 171182 dòng dữ liệu. Các cột `At least one dose_cumulative`, `Second_dose_cumulative`, `fully_vaccinated_cumulative` và `third_dose_cumulative` còn khuyết dữ liệu.

### Kiểm tra dữ liệu bị duplicated và xử lí

In [191]:
# Kiểm tra các dòng bị trùng lắp
vaccines_df[vaccines_df.duplicated()]

Empty DataFrame
Columns: [DATE, PHU_ID, AGEGROUP, AT_LEAST_ONE_DOSE_CUMULATIVE, SECOND_DOSE_CUMULATIVE, FULLY_VACCINATED_CUMULATIVE, THIRD_DOSE_CUMULATIVE]
Index: []

$\rightarrow$ **Kết quả:** Không có dòng dữ liệu bị duplicated nên không cần xử lí.

### Xuất dữ liệu hoàn chỉnh ra file source

In [192]:
vaccines_df.to_csv('Datasource_Cleaned/Vaccines.csv')